internal/pkg/instrumentation/bpf/google.golang.org/grpc/server/probe.go

```diff
@@ -15,205 +15,123 @@
+			},
+			probe.StructFieldConst{
+				Key: "frame_fields_pos",
+				Val: structfield.NewID("golang.org/x/net", "golang.org/x/net/http2", "MetaHeadersFrame", "Fields"),
+			},
+			probe.StructFieldConst{
+				Key: "frame_stream_id_pod",
+				Val: structfield.NewID("golang.org/x/net", "golang.org/x/net/http2", "FrameHeader", "StreamID"),
+			},
+		},
+		Uprobes: map[string]probe.UprobeFunc[bpfObjects]{
+			"google.golang.org/grpc.(*Server).handleStream":                           uprobeHandleStream,
+			"google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders": uprobeOperateHeaders,
+		},
 
-	g.bpfObjects = &bpfObjects{}
-	err = utils.LoadEBPFObjects(spec, g.bpfObjects, &ebpf.CollectionOptions{
-		Maps: ebpf.MapOptions{
-			PinPath: bpffs.PathForTargetApplication(target),
+		ReaderFn: func(obj bpfObjects) (*perf.Reader, error) {
+			return perf.NewReader(obj.Events, os.Getpagesize())
 		},
-	})
-	if err != nil {
-		return err
+		SpecFn:    loadBpf,
+		ProcessFn: convertEvent,
 	}
+}
 
-	offset, err := target.GetFunctionOffset(g.FuncNames()[0])
+func uprobeHandleStream(name string, exec *link.Executable, target *process.TargetDetails, obj *bpfObjects) ([]link.Link, error) {
+	offset, err := target.GetFunctionOffset(name)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	up, err := exec.Uprobe("", g.bpfObjects.UprobeServerHandleStream, &link.UprobeOptions{
-		Address: offset,
-	})
+	opts := &link.UprobeOptions{Address: offset}
+	l, err := exec.Uprobe("", obj.UprobeServerHandleStream, opts)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	links := []link.Link{l}
 
-	g.uprobe = up
-	retOffsets, err := target.GetFunctionReturns(g.FuncNames()[0])
+	retOffsets, err := target.GetFunctionReturns(name)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	for _, ret := range retOffsets {
-		retProbe, err := exec.Uprobe("", g.bpfObjects.UprobeServerHandleStreamReturns, &link.UprobeOptions{
-			Address: ret,
-		})
+		opts := &link.UprobeOptions{Address: ret}
+		l, err := exec.Uprobe("", obj.UprobeServerHandleStreamReturns, opts)
 		if err != nil {
-			return err
+			return nil, err
 		}
-		g.returnProbs = append(g.returnProbs, retProbe)
+		links = append(links, l)
 	}
 
-	headerOffset, err := target.GetFunctionOffset(g.FuncNames()[1])
-	if err != nil {
-		return err
-	}
-	hProbe, err := exec.Uprobe("", g.bpfObjects.UprobeDecodeStateDecodeHeader, &link.UprobeOptions{
-		Address: headerOffset,
-	})
+	return links, nil
+}
+
+func uprobeOperateHeaders(name string, exec *link.Executable, target *process.TargetDetails, obj *bpfObjects) ([]link.Link, error) {
+	offset, err := target.GetFunctionOffset(name)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	g.headersProbe = hProbe
 
-	rd, err := perf.NewReader(g.bpfObjects.Events, os.Getpagesize())
+	opts := &link.UprobeOptions{Address: offset}
+	l, err := exec.Uprobe("", obj.UprobeDecodeStateDecodeHeader, opts)
```

Comment:
It might be good to rename this BPF function in the BPF code, to match to the name of the function where the probe attaches to - `operateHeaders`. In older versions of gRPC the function used to be called `decodeHeader`, but that has been removed and the probe now attaches to `operateHeaders`.

Reply:
Sounds good to me. I'll open an issue to track this as a follow up. :+1:

Reply:
#497